opensearch-project/OpenSearch-Dashboards
# [Workspace] Add workspace id in basePath

### Description

In order to let users switch to specific workspace when paste a deep link from other users, we need to make url stateful with workspace info. This PR mainly introduce the capability to keep the workspace Id in url consistent with the workspace id in memory so that user can feel free to copy the url and share to others.

You can find all the options and discussions from the RFC issue: #5243 .

### Issues Resolved

#6015 

## Screenshot
![20240307174817818](https://github.com/opensearch-project/OpenSearch-Dashboards/assets/13493605/e22e2203-7687-431a-82d5-170d410ea78a)

## Testing the changes

- Using the branch to bootstrap
- Enable workspace by adding `workspace.enabled` to opensearch_dashboards.yml file 
- Start OSD by using `yarn start --no-base-path` to make sure no random base path will be appended
- Visit the OpenSearch Dashboards under `foo` workspace: `http://localhost:5601/w/foo/app/dev_tools`
- You will find `/w/foo` params will always be there when you navigate between left navigation, making the url shareable.

### Check List
